wandb/sdk/service/server_sock.py

```diff
@@ -0,0 +1,257 @@
+import queue
+import socket
+import threading
+import time
+from typing import Any, Callable, Dict, Optional
+from typing import TYPE_CHECKING
+
+from wandb.proto import wandb_server_pb2 as spb
+
+from .streams import _dict_from_pbmap
+from .streams import StreamMux
+from ..lib.sock_client import SockClient
+
+
+if TYPE_CHECKING:
+    from threading import Event
+    from ..interface.interface_relay import InterfaceRelay
+
+
+class ClientDict:
+    _client_dict: Dict[str, SockClient]
+    _lock: threading.Lock
+
+    def __init__(self) -> None:
+        self._client_dict = {}
+        self._lock = threading.Lock()
+
+    def get_client(self, client_id: str) -> Optional[SockClient]:
+        with self._lock:
+            client = self._client_dict.get(client_id)
+        return client
+
+    def add_client(self, client: SockClient) -> None:
+        with self._lock:
+            self._client_dict[client._sockid] = client
```

Comment:
can there be a case where are 2 clients with same `_sockid`? do we check for it?

Reply:
No case should exist, we are using uuid.uuid()
technically it could happen because only python3.7 support safe uuids...
ill add a comment about using safe uuids 

```
     def __init__(self) -> None:
         self._data = b""
+        # TODO: use safe uuid's (python3.7+) or emulate this
         self._sockid = uuid.uuid4().hex
```